python/ray/util/sgd/torch/examples/pytorch-lightning/mnist-ptl.py

```diff
@@ -0,0 +1,145 @@
+import argparse
+
+import torch
+from ray.util.sgd import TorchTrainer
+from ray.util.sgd.torch import TrainingOperator
+from torch.nn import functional as F
+from pytorch_lightning.core.lightning import LightningModule
+from torch.optim import Adam
+from torch.utils.data import DataLoader, random_split
+from torchvision.datasets import MNIST
+import os
+from torchvision import transforms
+
+
+class LitMNIST(LightningModule):
+    def __init__(self, config):
+        super().__init__()
+
+        # mnist images are (1, 28, 28) (channels, width, height)
+        self.layer_1 = torch.nn.Linear(28 * 28, 128)
+        self.layer_2 = torch.nn.Linear(128, 256)
+        self.layer_3 = torch.nn.Linear(256, 10)
+
+        self.config = config
+
+    def forward(self, x):
+        batch_size, channels, width, height = x.size()
+
+        # (b, 1, 28, 28) -> (b, 1*28*28)
+        x = x.view(batch_size, -1)
+        x = self.layer_1(x)
+        x = torch.relu(x)
+        x = self.layer_2(x)
+        x = torch.relu(x)
+        x = self.layer_3(x)
+
+        x = torch.log_softmax(x, dim=1)
+        return x
+
+    def configure_optimizers(self):
+        return Adam(self.parameters(), lr=self.config["lr"])
+
+    def setup(self, stage):
+        # transforms for images
+        transform = transforms.Compose([
+            transforms.ToTensor(),
+            transforms.Normalize((0.1307, ), (0.3081, ))
+        ])
+
+        # prepare transforms standard to MNIST
+        mnist_train = MNIST(
+            os.getcwd(), train=True, download=True, transform=transform)
+
+        self.mnist_train, self.mnist_val = random_split(
+            mnist_train, [55000, 5000])
+
+    def train_dataloader(self):
+        return DataLoader(
+            self.mnist_train, batch_size=self.config["batch_size"])
+
+    def val_dataloader(self):
+        return DataLoader(self.mnist_val, batch_size=self.config["batch_size"])
+
+    def training_step(self, batch, batch_idx):
+        x, y = batch
+        logits = self(x)
+        loss = F.nll_loss(logits, y)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        x, y = batch
+        logits = self(x)
+        loss = F.nll_loss(logits, y)
+        _, predicted = torch.max(logits.data, 1)
+        num_correct = (predicted == y).sum().item()
+        num_samples = y.size(0)
+        return {"val_loss": loss.item(), "val_acc": num_correct / num_samples}
+
+
+def train_mnist(num_workers=1, use_gpu=False, num_epochs=5):
+    Operator = TrainingOperator.from_ptl(LitMNIST)
```

Comment:
I'm thinking in the future we will want to support:

```
operator = TrainingOperator(LitMNIST())
trainer = TorchTrainer(training_operator=operator)
```
This will require some intense serialization tricks but it will be definitely useful (gets rid of the config).

Reply:
we would still need to take in a config if we want to use this with tune right?